src/coreclr/jit/lsraxarch.cpp

```diff
@@ -2328,48 +2328,77 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
 
                 const bool copiesUpperBits = HWIntrinsicInfo::CopiesUpperBits(intrinsicId);
 
-                // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
-                assert(!copiesUpperBits || !op1->isContained());
+                unsigned resultOpNum = 0;
+                LIR::Use use;
+                GenTree* user = nullptr;
 
-                if (op2->isContained())
+                if (LIR::AsRange(blockSequence[curBBSeqNum]).TryGetUse(intrinsicTree, &use))
                 {
-                    // 132 form: op1 = (op1 * op3) + [op2]
+                    user = use.User();
+                }
+                resultOpNum = intrinsicTree->GetResultOpNumForFMA(user, op1, op2, op3);
 
-                    tgtPrefUse = BuildUse(op1);
+                // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
+                assert(!copiesUpperBits || !op1->isContained());
 
-                    srcCount += 1;
-                    srcCount += BuildOperandUses(op2);
-                    srcCount += BuildDelayFreeUses(op3, op1);
+                unsigned containedOpNum = 0;
+
+                if (op1->isContained() || op1->IsRegOptional())
+                {
+                    containedOpNum = 1;
                 }
-                else if (op1->isContained())
+                else if (op2->isContained() || op2->IsRegOptional())
                 {
-                    // 231 form: op3 = (op2 * op3) + [op1]
-
-                    tgtPrefUse = BuildUse(op3);
-
-                    srcCount += BuildOperandUses(op1);
-                    srcCount += BuildDelayFreeUses(op2, op1);
-                    srcCount += 1;
+                    containedOpNum = 2;
                 }
                 else
                 {
-                    // 213 form: op1 = (op2 * op1) + [op3]
+                    assert(op3->isContained() || op3->IsRegOptional());
+                    containedOpNum = 3;
+                }
 
-                    tgtPrefUse = BuildUse(op1);
-                    srcCount += 1;
+                GenTree* emitOp1 = op1;
+                GenTree* emitOp2 = op2;
+                GenTree* emitOp3 = op3;
 
-                    if (copiesUpperBits)
+                // Intrinsics with CopyUpperBits semantics must have op1 as target
+                if (containedOpNum == 1 && !copiesUpperBits)
+                {
+                    if (resultOpNum != 3)
```

Comment:
It would be nice if this were the positive case and there were an assert that `resultOpNum != containedOpNum`

Therefore, if `containedOpNum == 1` then `resultOpNum` can only be `0`, `2`, or `3`

If it's `3`, then swapping `op1`/`op3` is sufficient
If it's `2`, then swapping `op2`/`op3` is needed first
If it's `0`, then it doesn't matter what we do so its fine to not swap

Reply:
is it possible that ```containedOpNum ==0``` and ```resultOpNum==0```?

Reply:
If none of the operands are overwritten and none are last use, then `containedOpNum == 0`.

I think we probably won't also get `containedOpNum == 0` because `VFMADD` should support general-purpose loads as well and so `RegOptional` should probably be true for at least one case. But in general its better to check and account for possible future changes, scenarios, or nodes that are introduced

Reply:
Because ```op lastUse``` could be updated after lowering, there are cases that we have ```resultOpNum == containedOpNum``` when they are not ```0```.

Reply:
Can we make multiple ops contained in lowering or change that in lsra?